File: TS/src/types/video.ts
// Types - Video

import { Format, HasVideo, HasAudio, IsAudioOnly, QualityRank } from "./format.js";
import { Thumbnail } from "./thumbnail.js";

export interface Video {
    ID: string;
    Title: string;
    Description: string;
    Author: string;
    ChannelID: string;
    ChannelURL: string;
    Duration: number;
    ViewCount: number;
    PublishDate: string;
    UploadDate: string;
    IsLive: boolean;
    IsLiveContent: boolean;
    IsPlayable: boolean;
    IsPrivate: boolean;
    IsUpcoming: boolean;
    LiveBroadcastDetails: LiveBroadcastDetails | null;
    Thumbnails: Thumbnail[];
    Formats: Format[];
    DASHManifestURL: string;
    HLSManifestURL: string;
    PlayabilityStatus: PlayabilityStatus;
}

export interface PlayabilityStatus {
    Status: string;
    Reason: string;
    PlayableInEmbed: boolean;
    IsAgeRestricted: boolean;
    MiniplayerStatus: string;
}

export interface LiveBroadcastDetails {
    IsLiveNow: boolean;
    StartTimestamp: string;
    EndTimestamp: string;
}

// FilterFormats returns formats that match the predicate
export function FilterFormats(v: Video, predicate: (f: Format) => boolean): Format[] {
    return v.Formats.filter(predicate);
}

// VideoFormats returns all formats that contain video
export function VideoFormats(v: Video): Format[] {
    return FilterFormats(v, (f) => HasVideo(f));
}

// AudioFormats returns all formats that contain audio only
export function AudioFormats(v: Video): Format[] {
    return FilterFormats(v, (f) => IsAudioOnly(f));
}

// BestVideoFormat returns the best video format (highest resolution, then bitrate)
export function BestVideoFormat(v: Video): Format | null {
    const formats = VideoFormats(v);
    if (formats.length === 0) {
        return null;
    }

    return formats.reduce((best, current) => {
        // First compare by resolution
        const currentPixels = current.Width * current.Height;
        const bestPixels = best.Width * best.Height;

        if (currentPixels !== bestPixels) {
            return currentPixels > bestPixels ? current : best;
        }

        // Then by bitrate
        if (current.Bitrate !== best.Bitrate) {
            return current.Bitrate > best.Bitrate ? current : best;
        }

        return best;
    });
}

// BestAudioFormat returns the best audio format (highest bitrate)
export function BestAudioFormat(v: Video): Format | null {
    const formats = AudioFormats(v);
    if (formats.length === 0) {
        return null;
    }

    return formats.reduce((best, current) => {
        // Compare by audio quality rank first
        const currentRank = QualityRank(current.AudioQuality);
        const bestRank = QualityRank(best.AudioQuality);

        if (currentRank !== bestRank) {
            return currentRank > bestRank ? current : best;
        }

        // Then by bitrate
        if (current.Bitrate !== best.Bitrate) {
            return current.Bitrate > best.Bitrate ? current : best;
        }

        return best;
    });
}

// GetFormat returns the format with the given itag
export function GetFormat(v: Video, itag: number): Format | null {
    return v.Formats.find((f) => f.ITag === itag) || null;
}

// HasLiveFormats returns true if the video has live streaming formats (HLS/DASH)
export function HasLiveFormats(v: Video): boolean {
    return v.DASHManifestURL !== "" || v.HLSManifestURL !== "";
}

// GetThumbnail returns the best thumbnail (highest resolution)
export function GetThumbnail(v: Video): Thumbnail | null {
    if (v.Thumbnails.length === 0) {
        return null;
    }

    return v.Thumbnails.reduce((best, current) => {
        const currentPixels = current.Width * current.Height;
        const bestPixels = best.Width * best.Height;
        return currentPixels > bestPixels ? current : best;
    });
}

// DurationString returns the duration as a human-readable string (HH:MM:SS)
export function DurationString(v: Video): string {
    const hours = Math.floor(v.Duration / 3600);
    const minutes = Math.floor((v.Duration % 3600) / 60);
    const seconds = v.Duration % 60;

    if (hours > 0) {
        return `${hours}:${minutes.toString().padStart(2, "0")}:${seconds.toString().padStart(2, "0")}`;
    }
    return `${minutes}:${seconds.toString().padStart(2, "0")}`;
}

// CreateEmptyVideo creates an empty Video with default values
export function CreateEmptyVideo(): Video {
    return {
        ID: "",
        Title: "",
        Description: "",
        Author: "",
        ChannelID: "",
        ChannelURL: "",
        Duration: 0,
        ViewCount: 0,
        PublishDate: "",
        UploadDate: "",
        IsLive: false,
        IsLiveContent: false,
        IsPlayable: false,
        IsPrivate: false,
        IsUpcoming: false,
        LiveBroadcastDetails: null,
        Thumbnails: [],
        Formats: [],
        DASHManifestURL: "",
        HLSManifestURL: "",
        PlayabilityStatus: {
            Status: "",
            Reason: "",
            PlayableInEmbed: false,
            IsAgeRestricted: false,
            MiniplayerStatus: "",
        },
    };
}

File: V2/types/video.go
// Package types provides core type definitions for the YouTube library.
package types

import "fmt"

// Video represents a YouTube video with all its metadata and available formats
type Video struct {
	ID          string
	Title       string
	Description string
	Duration    int
	ViewCount   int

	Author      string
	ChannelID   string
	ChannelURL  string
	UploadDate  string

	Thumbnails []Thumbnail
	Keywords   []string
	Category   string

	IsLive        bool
	IsLiveContent bool
	IsPrivate     bool
	AgeRestricted bool

	Formats []Format

	// Internal data for format URL resolution
	VisitorData string
	DataSyncID  string
	PlayerURL   string
}

// VideoDetails provides a simplified view of video metadata
type VideoDetails struct {
	ID          string
	URL         string
	Title       string
	Description string
	Duration    int
	ViewCount   int

	Author    string
	ChannelID string

	Thumbnails []Thumbnail
	Keywords   []string

	IsLive        bool
	IsPrivate     bool
	AgeRestricted bool
}

// URL returns the video watch URL
func (v *Video) URL() string {
	return fmt.Sprintf("https://www.youtube.com/watch?v=%s", v.ID)
}

// Details returns a simplified VideoDetails struct
func (v *Video) Details() VideoDetails {
	return VideoDetails{
		ID:            v.ID,
		URL:           v.URL(),
		Title:         v.Title,
		Description:   v.Description,
		Duration:      v.Duration,
		ViewCount:     v.ViewCount,
		Author:        v.Author,
		ChannelID:     v.ChannelID,
		Thumbnails:    v.Thumbnails,
		Keywords:      v.Keywords,
		IsLive:        v.IsLive,
		IsPrivate:     v.IsPrivate,
		AgeRestricted: v.AgeRestricted,
	}
}

// BestThumbnail returns the highest resolution thumbnail
func (v *Video) BestThumbnail() *Thumbnail {
	if len(v.Thumbnails) == 0 {
		return nil
	}

	best := &v.Thumbnails[0]
	for i := range v.Thumbnails {
		if v.Thumbnails[i].Width > best.Width {
			best = &v.Thumbnails[i]
		}
	}

	return best
}

// FilterFormats returns formats matching the given filter function
func (v *Video) FilterFormats(filter func(f Format) bool) []Format {
	var result []Format
	for _, f := range v.Formats {
		if filter(f) {
			result = append(result, f)
		}
	}
	return result
}

// VideoFormats returns only formats with video
func (v *Video) VideoFormats() []Format {
	return v.FilterFormats(func(f Format) bool {
		return f.HasVideo()
	})
}

// AudioFormats returns only formats with audio (including audio-only)
func (v *Video) AudioFormats() []Format {
	return v.FilterFormats(func(f Format) bool {
		return f.HasAudio()
	})
}

// AudioOnlyFormats returns only audio-only formats
func (v *Video) AudioOnlyFormats() []Format {
	return v.FilterFormats(func(f Format) bool {
		return f.IsAudioOnly()
	})
}

// VideoOnlyFormats returns only video-only formats (no audio)
func (v *Video) VideoOnlyFormats() []Format {
	return v.FilterFormats(func(f Format) bool {
		return f.IsVideoOnly()
	})
}

// StreamableFormats returns formats that support range requests
func (v *Video) StreamableFormats() []Format {
	return v.FilterFormats(func(f Format) bool {
		return f.ContentLength > 0
	})
}
